### PR TITLE
feat: add version and party filters

### LINKED ISSUE #44: Add a filter on type '1st pary' or '3rd-party' plugin
Hello,

This could help users to find plugins from Adonis.

### PR DESCRIPTION
**Pull request by Barbapapazes**
fix #44

This PR:

- refactor how the home page handle filters, order, orderby and search
- rename sort to orderBy and introduce order
- introduce version and party filters

### CODE REVIEW

**Comment by Julien-R44** [comment]
Woooow that's cool thanks bro
i would like to refactor some small parts of the code 
same for the UI, there are a few places I'd like to rework. In particular the responsive ( below the desktop format ) which I think could be improved quite a bit. If you feel like doing it, i would love it

otherwise it's very cool thanks. will do a more complete review in the next few days.

**Comment by Barbapapazes** [comment]
> Woooow that's cool thanks bro i would like to refactor some small parts of the code same for the UI, there are a few places I'd like to rework. In particular the responsive ( below the desktop format ) which I think could be improved quite a bit. If you feel like doing it, i would love it
> 
> otherwise it's very cool thanks. will do a more complete review in the next few days.

Yes, the mobile version need to be polished. I've done it quickly at the end. I'm waiting a more in-depth review to update this PR.

**Comment by Julien-R44** [comment]
Refactored some things : 

- Basically used defineModel where possible. This is more concise
- Refactored main.vue to make it a bit cleaner and avoid some duplication

Let me know if something looks a bit off to you. 
Also, if you are up to working on the responsive version that would be with pleasure

File: app/services/packages_fetcher.ts [hunk 1]
```diff
@@ -93,6 +110,33 @@ export class PackagesFetcher {
...
     }
 
+    if (options.parties) {
+      packages = packages.filter((pkg) => {
+        return options.parties!.includes(pkg.type)
+      })
+    }
+
+    if (options.versions) {
+      const specifiersRegex = /[\^~]/g
```
Comment (Julien-R44) [inline app/services/packages_fetcher.ts:120]: I think we can extract this into a `filterVersionPackages` method in the same class

**Comment by Barbapapazes** [comment]
> Refactored some things :
> 
> * Basically used defineModel where possible. This is more concise
> * Refactored main.vue to make it a bit cleaner and avoid some duplication
> 
> Let me know if something looks a bit off to you. Also, if you are up to working on the responsive version that would be with pleasure

Nice improvements and refactor! Really like them (reduce code and improve readability)

Will work on the responsive version.

**Comment by Julien-R44** [comment]
Also, maybe it's overkill. But I wonder if it wouldn't be a good idea to extract components like `<SelectPartiesMenu />` `<SelectVersionMenu>` to clean up `main.vue` a bit more. Or we could probably just extract the `partiesOptions` `versionOptions` and `ordersByOptions` etc. in a separate file. Or just leave it as it is, its fine too

File: app/services/packages_fetcher.ts (hunk 1 above)
Comment (Barbapapazes) [reply app/services/packages_fetcher.ts:120]: We can do it but don't we have to do the same for the `party` and `seach` option?

**Comment by Barbapapazes** [comment]
> Also, maybe it's overkill. But I wonder if it wouldn't be a good idea to extract components like `<SelectPartiesMenu />` `<SelectVersionMenu>` to clean up `main.vue` a bit more. Or we could probably just extract the `partiesOptions` `versionOptions` and `ordersByOptions` etc. in a separate file. Or just leave it as it is, its fine too

Will take a look when working on the responsive part (because I'm thinking of way to also improve the UX).

File: app/services/packages_fetcher.ts (hunk 1 above)
Comment (Julien-R44) [reply app/services/packages_fetcher.ts:120]: Yeah, but the logic for filtering by version is rather low-level and verbose. Having a : 

```ts
    if (options.category) {
      packages = packages.filter((pkg) => pkg.category === options.category)
    }
```

doesn't particularly shock me, for example, it's just a single line of code, without much logic. But for consistency's sake, we could also do it, yes, that would be cool

In any case, don't worry too much about it. Right now I have to work on something else, but I'll come back to it this weekend or tonight and I'll be able to take care of it

**Comment by Barbapapazes** [comment]
Just push some update for the mobile version. Let me know what you think and do no hesitate to update style.

**Comment by Julien-R44** [comment]
thanks dude!
I've noticed a few things: 

- when you change the filters, they are saved in the route query, but when you refresh the page they are not reused which causes a little weird behavior

- Responsive. I think we can probably still improve on that. example on 13". 
image

but I don't really have any good suggestions for that. that's the main problem I had when I was thinking about it too. That said, I may have an idea that requires a bit of work: 
Rather than having 2 different selectors for versions and parties, maybe we could have just one? So we'd have a single dropdown called "selected filters" and that would open a single dropdown with all filters, with a separator between "versions" and "parties" ? Maybe that's a bad idea. i am not sure.

- Filter modal. It looks very rough to me. I think we could improve the UI a bit: have real buttons with background color or borders. The same goes for dropdowns, maybe a backround color to distinguish dropdowns from modals? 
image

I would have liked to do it myself so as not to bother you, but I won't have time to work on it next week unfortunately. big week at work incoming 😅

**Comment by Julien-R44** [comment]
finally found some time to work on it! I've updated the UI a bit :

image

also fixed the last little bugs

Let me know if it looks okay to you. Gonna merge this but feel free to re-open an issue if needed
Thanks a lot for taking time to work on that with me! Was cool 🤝

**Pull request merged**